src/NHibernate/Linq/Visitors/NullableExpressionDetector.cs

```diff
@@ -0,0 +1,309 @@
+					return true;
+			}
+		}
+
+		private bool IsNullable(MemberExpression memberExpression, BinaryExpression equalityExpression)
+		{
+			if (_functionRegistry.TryGetGenerator(memberExpression.Member, out _))
+			{
+				// We have to skip the property as it will be converted to a function that can return null
+				// if the argument is null
+				return IsNullable(memberExpression.Expression, equalityExpression);
+			}
+
+			var memberType = memberExpression.Member.GetPropertyOrFieldType();
+			if (memberType?.IsValueType == true && !memberType.IsNullable())
+			{
+				return IsNullable(memberExpression.Expression, equalityExpression);
+			}
+
+			// Check if there was a not null check prior the equality expression
+			if ((
+				    equalityExpression.NodeType == ExpressionType.NotEqual ||
+				    equalityExpression.NodeType == ExpressionType.Equal
+			    ) &&
+			    _equalityNotNullMembers.TryGetValue(equalityExpression, out var notNullMembers) &&
+			    notNullMembers.Any(o => AreEqual(o, memberExpression)))
+			{
+				return false;
+			}
+
+			// We have to check the member mapping to determine if is nullable
+			var entityName = TryGetEntityName(memberExpression);
+			if (entityName == null)
+			{
+				return true; // not mapped
+			}
+
+			var persister = _sessionFactory.GetEntityPersister(entityName);
+			var index = persister.EntityMetamodel.GetPropertyIndexOrNull(memberExpression.Member.Name);
+			if (!index.HasValue || persister.EntityMetamodel.PropertyNullability[index.Value])
+			{
+				return true; // not mapped or nullable
+			}
+
+			return IsNullable(memberExpression.Expression, equalityExpression);
+		}
+
+		private bool IsNullableExtension(Expression extensionExpression, BinaryExpression equalityExpression)
+		{
+			switch (extensionExpression)
+			{
+				case QuerySourceReferenceExpression querySourceReferenceExpression:
+					switch (querySourceReferenceExpression.ReferencedQuerySource)
+					{
+						case MainFromClause _:
+							return false; // we reached to the root expression, there were no nullable expressions
+						case NhJoinClause joinClause:
+							return IsNullable(joinClause.FromExpression, equalityExpression);
+						default:
+							return true; // unknown query source
+					}
+				case SubQueryExpression subQueryExpression:
+					if (subQueryExpression.QueryModel.SelectClause.Selector is NhAggregatedExpression subQueryAggregatedExpression)
+					{
+						return subQueryAggregatedExpression.AllowsNullableReturnType;
+					}
+					else if (subQueryExpression.QueryModel.ResultOperators.Any(o => NotNullOperators.Contains(o.GetType())))
+					{
+						return false;
+					}
+
+					return true;
+				case NhAggregatedExpression aggregatedExpression:
+					return aggregatedExpression.AllowsNullableReturnType;
+				default:
+					return true; // a query can return null and we cannot calculate it as it is not yet executed
+			}
+		}
+
+		private string TryGetEntityName(MemberExpression memberExpression)
```

Comment:
After doing #2079 I found out that this method does not cover all cases. I will update this method and try to add additional tests that cover the edge cases.